### ISSUE TITLE
Need Chef::EncryptedAttribute.exist_on_node? method (and RemoteNode class) to use knife search query keys rather than only node's name 

### ISSUE THREAD

**Issue by chhsiung** [onddo/chef-encrypted-attributes#3]
I'm running into a problem where Chef::EncryptedAttribute.exist_on_node?(name-of-my-node) is returning multiple nodes due to the fact that I have metadata on other nodes with the same 'name = name-of-my-node' data. In our current Chef environment, we believe this is perfectly acceptable and necessary data. An example of this occurring is the metadata a proxmox server may need. According to the knife search documentation, 'When nested fields are present in a JSON structure, the chef-client will extract those nested fields to the top-level, flattening them into compound fields that support wildcard search patterns.' which means Chef's search is working properly given the search criteria.

I'm logging this as an issue first before proposing a pull request because I'm not sure how to fix this without breaking backwards compatibility. I think the way chef-encrypted-attribute currently searches for the nodes to encrypt for (using the usual search query keys rather than hardcoding a specific key to search on) seems a reasonable way to find the node that the encrypted attribute exists on. But I'm not sure how to make this change and not break backwards compatibility.

Finally, is there any other way around this? Am I not using Chef::EncryptedAttribute.exist_on_node? properly?

Thanks

**Comment by zuazo** [comment]
Thanks @chhsiung. Interesting bug. I think I understand your problem. You have some nodes with attributes similar to the following:

```json
{
  "name": "cat1.example.com",
  "run_list": [ "role[meow]" ],
  "normal": {
    "proxmox_servers": {
      "web": {
        "name": "mouse001.example.com",
        "ip": "192.168.0.25"
      }
    }
  }
}
```

And using `$ knife search node "name:mouse001.example.com"` also returns this server as a result. The same applies to `.exist_on_node?` internal search and everything fails.

I think this error can be fixed only by checking the server that returned search result. In the normal search the server name is returned by default and I think that in partial search you can use the `"name": ["name"]` key search to return the server name. But I'm not sure.

I will try to fix this, but give me a few days because I'm a bit busy right now. If you're in a hurry and you're thinking about making a PR, remember to include tests and to update the documentation accordingly. I think this could require a big change in the `SearchHelper` class (a *name* parameter?) but maybe not so complicated.

**Comment by chhsiung** [comment]
@zuazo - any update on this? sorry, i hadn't had a chance to take a look at this since i logged the issue so haven't tried your suggested fix yet.

**Comment by zuazo** [comment]
@chhsiung, my apologies, I have to fixed this yet. Some tests with chef-zero suggest that can be fixed as I said. But I have not had the time to fix it yet. I will try to spend some time on it this weekend.

**Comment by chhsiung** [comment]
Hi @zuazo, sorry haven't had a chance to revisit this earlier, but I finally played with the partial_search you suggested!  I'm not sure whether I'm doing things wrong.... but when I pass in "name"=>["name"] to the partial search, i am still receiving back extra nodes, but just a hash of their values now.

(so assuming 2 nodes, one called "secret-keeper" and another having other attributes with "name":"secret-keeper")

before:
[node[secret-keeper], node[not-secret-keeper]]  

now, with partial search:
[{"name"=>"secret-keeper"}, {"name"=>"not-secret-keeper"}] 

am i using partial search correctly? if so, should i select 1 more time on the results to make sure i retrieve the correct node?

**Comment by zuazo** [comment]
Hi @chhsiung,

I fixed this in master. Including tests and documentation. I hope I did not break anything.

Doing partial search with `name: ["name"]` still returns the extra nodes, but you can filter them now knowing their real node name and get only the correct one.

**Comment by chhsiung** [comment]
@zuazo - thanks so much for your quick fix!  I played around with it and it appears to be working for me in test-kitchen, returning just 1 node and the proper one! thanks again! :)

**Comment by zuazo** [comment]
Crystal, thanks for the feedback. I released gem version [`0.7.0`](https://github.com/onddo/chef-encrypted-attributes/releases/tag/0.7.0) with the fix.

**Comment by chhsiung** [comment]
@zuazo, hi. again. :(

things were working great in test-kitchen! i had even copied over the production nodes' data in order to recreate the same exact scenario.  but then things didn't work for me in production once the new 0.7.0 gem was released (btw, no problems installing 0.7.0, thanks for that fix!!)

i *think* the problem is related to https://github.com/onddo/chef-encrypted-attributes/blob/master/lib/chef/encrypted_attribute/remote_node.rb#L79. In this call, the # rows is set to 1... Since that row value gets passed all the way through to the Chef search call (https://github.com/onddo/chef-encrypted-attributes/blob/master/lib/chef/encrypted_attribute/search_helper.rb#L372), only 1 row is ever returned (which in production happens to NOT be the one i want; i guess in test-kitchen, I had created the nodes in a different order) and filter_normal_search_response never gets to actually filter through multiple results.

i tested this theory out by upping the row parameter to 2 and i was able to successfully retrieve the attribute from the right node.  obviously don't want to leave it at 2, but i don't want to not set a limit... what do you suggest?

**Comment by zuazo** [comment]
Thanks @chhsiung for all your patience and feedback. I increased the row limit to `1000` and added the `search_max_rows` configuration option to increase this limit if necessary.

**Comment by zuazo** [comment]
Another release with the last fix: [`0.8.0`](https://github.com/onddo/chef-encrypted-attributes/releases/tag/0.8.0). Let me know if it does not work for you :wink:
